File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head is None:
            return None
        
        slow = head
        fast = head
        flag = False
        # Detect cycle using slow at 1x and fast at 2x speed 
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

            if slow == fast:
                flag = True
                break

        if not flag:
            return None
        #  reset slow to head
        slow = head
#  where ever the slow and fast meets that's the node where cycle begins.
        while slow != fast:
            slow = slow.next
            fast = fast.next
        
        return slow

# Time Complexity: O(n)
# Space Complexity: O(1)

File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.java
// /
//   Definition for singly-linked list.
//   struct ListNode {
//       int val;
//       ListNode next;
//       ListNode(int x) : val(x), next(NULL) {}
//   };
//  /
class Solution {
    public ListNode detectCycle(ListNode head) {
        // if(head == null){
        //     return None;
        // }

        ListNode slow = head;
        ListNode fast = head;
        boolean flag = false;

        //detect cycle
        while(fast != null && fast.next != null){
            slow = slow.next;
            fast = fast.next.next;

            if(slow == fast){
                flag = true;
                break;                    
            }
        }

        //if no cycle then return null
        if(flag == false){
            return null;
        }

        slow = head;

        //whereever the slow and fast meets that's the node where cycle begins 
        while(slow != fast){
            slow = slow.next;
            fast = fast.next;
        }
        return slow; 
    }
};